File: infoset.py
import numpy as np

class Infoset:

    def __init__(self, num_actions):
        self.cumul_regret = np.zeros(num_actions)
        self.cumul_strategy = np.zeros(num_actions)

    def get_strategy(self):
        """
        Gets current information set mixed strategy through regret-matching.
        """
        strategy = self.cumul_regret.clip(0)
        normalizing_sum = np.sum(strategy)
        if normalizing_sum > 0:
            return strategy / normalizing_sum
        return np.full_like(strategy, 1.0 / len(strategy))

    def get_average_strategy(self):
        """
        Gets average information set mixed strategy across all training iterations.
        """
        normalizing_sum = np.sum(self.cumul_strategy)
        if normalizing_sum > 0:
            return self.cumul_strategy / normalizing_sum
        return np.full_like(self.cumul_strategy, 1.0 / len(self.cumul_strategy))

File: mccfr/Infoset.java
package mccfr;

import java.io.Serializable;

public class Infoset implements Serializable {
    public float[] cumulRegret;
    public float[] cumulStrategy;

    public Infoset(int numActions) {
        cumulRegret = new float[numActions];
        cumulStrategy = new float[numActions];
    }

    public float[] getStrategy() {
        float normalizingSum = 0;
        float[] strategy = new float[cumulRegret.length];
        for (int i = 0; i < cumulRegret.length; i++) {
            normalizingSum += Math.max(0, cumulRegret[i]);
        }
        for (int i = 0; i < cumulRegret.length; i++) {
            if (normalizingSum > 0) {
                strategy[i] = Math.max(0, cumulRegret[i] / normalizingSum);
            } else {
                strategy[i] = 1.0f / cumulRegret.length;
            }
        }
        return strategy;
    }

    public float[] getAverageStrategy() {
        float normalizingSum = 0;
        float[] averageStrategy = new float[cumulStrategy.length];
        for (int i = 0; i < cumulStrategy.length; i++) {
            normalizingSum += cumulStrategy[i];
        }
        for (int i = 0; i < cumulStrategy.length; i++) {
            if (normalizingSum > 0) {
                averageStrategy[i] = cumulStrategy[i] / normalizingSum;
            } else {
                averageStrategy[i] = 1.0f / cumulStrategy.length;
            }
        }
        return averageStrategy;
    }
}
